src/Implementations/BaseMessageBus.ts

```diff
@@ -0,0 +1,107 @@
+import {HandlersMap} from '../Lib/HandlersMap';
+import {Message} from '../Lib/Message';
+import {MessageBus} from '../Lib/MessageBus';
+import {MessageBusMiddleware} from '../Lib/MessageBusMiddleware';
+import {MessageHandler} from '../Lib/MessageHandler';
+import {MessageHandlerObject} from '../Lib/MessageHandlerObject';
+import {isMessage} from './TypeCheckers/isMessage';
+import {isMessageHandler} from './TypeCheckers/isMessageHandler';
+import {isMessageHandlerFunction} from './TypeCheckers/isMessageHandlerFunction';
+import {isMessageHandlerObject} from './TypeCheckers/isMessageHandlerObject';
+
+export abstract class BaseMessageBus implements MessageBus {
+    protected handlersMap: HandlersMap = {};
+    protected middlewareList: MessageBusMiddleware<any>[] = [];
+
+    async handle<P>(message: Message<P>): Promise<void> {
+        const type = message.type;
+
+        for (const middleware of this.middlewareList) {
+            // We're not catching potential exceptions.
+            // One middleware crashes, the message won't even reach the handlers.
+            await middleware.handle(message);
+        }
+
+        if (!this.handlersMap[type]) {
```

Comment:
If you use a `Map` class in `handlersMap` you can assign handlers to const and check this condition.